### PR TITLE
Support in-place upgrades of new minio releases

### LINKED ISSUE #4781: Using go-update explore/implement a self-updating minio server
> # go-update: Build self-updating Go programs
>Package update provides functionality to implement secure, self-updating Go programs (or other single-file targets) A program can update itself by replacing its executable file with a new version.
>It provides the flexibility to implement different updating user experiences like auto-updating, or manual user-initiated updates. It also boasts advanced features like binary patching and code signing verification.

Use https://github.com/inconshreveable/go-update and explore the possibility of self updating minio server. 

A sample code is here 

```go
func doUpdateWithChecksum(checkSum string, latestReleaseTime time.Time) (string, error) {
        byteCksum, err := hex.DecodeString(checkSum)
        if err != nil {
                return "", err
        }
        resp, err := http.Get("https://dl.minio.io/server/minio/release/linux-amd64/minio")
        if err != nil {
                return "", err
        }
        defer resp.Body.Close()
        if err = update.RollbackError(update.Apply(resp.Body,
                update.Options{
                        Hash:     crypto.SHA256,
                        Checksum: byteCksum,
                },
        )); err != nil {
                return "", err
        }
        return fmt.Sprintf("Minio updated to version %s successfully.", releaseTime)
}
```

Implementation requires two things

 - `minio update` updates itself.
 - `mc admin update` updates minio server for an entire cluster.

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
This PR allows 'minio update' to not only shows update banner
but also allows for in-place upgrades.

Updates are done safely by validating the downloaded
sha256 of the binary.

## Motivation and Context
Fixes #4781

## How Has This Been Tested?
Manually using `minio update`

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by harshavardhana** [comment]
This PR is ready for review to test this patch out just do the following steps.

```
$ go build --ldflags "-s -w -X github.com/minio/minio/cmd.Version=2017-09-18T08:38:47Z -X github.com/minio/minio/cmd.ReleaseTag=RELEASE.2017-09-18T08-38-47Z -X github.com/minio/minio/cmd.CommitID=64bde1340cf3530decfca58059cb4196a273490b -X github.com/minio/minio/cmd.ShortCommitID=64bde1340cf3 -X github.com/minio/minio/cmd.GOPATH=/home/harsha/mygo"
```

```
$ ./minio server ~/test

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ You are running an older version of Minio released 1 week ago      ┃
┃ Update: https://dl.minio.io/server/minio/release/linux-amd64/minio ┃
┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
```

```
$ ./minio update

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ You are running an older version of Minio released 1 week ago      ┃
┃ Update: https://dl.minio.io/server/minio/release/linux-amd64/minio ┃
┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛

Update to RELEASE.2017-09-29T19-16-56Z [yes]: yes
Minio updated to version RELEASE.2017-09-29T19-16-56Z successfully.
```

```
$ ./minio version
Version: 2017-09-29T19:16:56Z
Release-Tag: RELEASE.2017-09-29T19-16-56Z
Commit-ID: 60cc6184d253efee4a3120683517028342229e21
```

**Comment by harshavardhana** [comment]
The UI needs discussion with @abperiasamy but PR can be reviewed for other things.

**Comment by harshavardhana** [comment]
One more additional thing to notice is

- Docker containers
- Kubernetes
- Mesos 

Are ignored for this update feature since each of them have their own update requirements.

**Comment by nitisht** [comment]
> The UI needs discussion with @abperiasamy but PR can be reviewed for other things.

Merging blocked till this happens, but code review can be done

**Comment by deekoder** [comment]
Is this now again ready for review?

**Comment by harshavardhana** [comment]
> Is this now again ready for review?

Its been ready for a while @deekoder - requires discussion with @abperiasamy as well.

**Comment by harshavardhana** [comment]
Discussed this with @abperiasamy  and we have a way of addressing the problems of update functionality on orchestration platforms. 

We will be adding a ENV such as MINIO_UPDATE which will be set to off on each orchestration platforms such that this functionality is disabled.

**Comment by harshavardhana** [comment]
This is approved by @abperiasamy from the UI perspective, fixing the remaining MINIO_UPDATE env for orchestration platforms to turn-off auto updates.

**Comment by harshavardhana** [comment]
The UI changes are following when server starts

```
$ ./minio server ~/test

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ You are running an older version of Minio released 2 months ago ┃
┃ Update: Run `minio update`                                      ┃
┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛

Endpoint:  http://192.168.1.17:9000  http://172.17.0.1:9000  http://127.0.0.1:9000
AccessKey: USWUXHGYZQYFYFFIT3RE 
SecretKey: MOJRH0mkL1IPauahWITSVvyDrQbEEIwljvmxdq03 
SQS ARNs:  arn:minio:sqs::test1:redis arn:minio:sqs::test2:redis

Browser Access:
   http://192.168.1.17:9000  http://172.17.0.1:9000  http://127.0.0.1:9000

Command-line Access: https://docs.minio.io/docs/minio-client-quickstart-guide
   $ mc config host add myminio http://192.168.1.17:9000 USWUXHGYZQYFYFFIT3RE MOJRH0mkL1IPauahWITSVvyDrQbEEIwljvmxdq03

Object API (Amazon S3 compatible):
   Go:         https://docs.minio.io/docs/golang-client-quickstart-guide
   Java:       https://docs.minio.io/docs/java-client-quickstart-guide
   Python:     https://docs.minio.io/docs/python-client-quickstart-guide
   JavaScript: https://docs.minio.io/docs/javascript-client-quickstart-guide
   .NET:       https://docs.minio.io/docs/dotnet-client-quickstart-guide

Drive Capacity: 71 GiB Free, 156 GiB Total
```

```
$ ./minio update

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ You are running an older version of Minio released 2 months ago    ┃
┃ Update: https://dl.minio.io/server/minio/release/linux-amd64/minio ┃
┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛

Update to RELEASE.2017-11-22T19-55-46Z [yes]: yes
Minio updated to version RELEASE.2017-11-22T19-55-46Z successfully.
```

If the MINIO_UPDATE=off is set then all the update functionality is disabled and kept as is with current master. 

This PR is ready for review and merge @nitisht @krisis

**Comment by harshavardhana** [comment]
Assigning review to @donatello - @krisis is unavailable.

File: cmd/gateway-sia.go [hunk 1]
```diff
@@ -63,6 +63,12 @@ ENVIRONMENT VARIABLES: (Default values in parenthesis)
      MINIO_ACCESS_KEY: Custom access key (Do not reuse same access keys on all instances)
      MINIO_SECRET_KEY: Custom secret key (Do not reuse same secret keys on all instances)
 
+  BROWSER:
+     MINIO_BROWSER: To disable web browser access, set this value to "off".
+
+  UPDATE:
+     MINIO_UPDATE: Turn off in-place upgrades, set this value to "off".
```
Comment (nitisht) [inline cmd/gateway-sia.go]: _To_ turn off in-place upgrades, set this value to "off". Same in other places

**Review by nitisht — CHANGES_REQUESTED** [review]

File: cmd/common-main.go [hunk 2]
```diff
@@ -29,10 +29,12 @@ import (
 
 // Check for updates and print a notification message
 func checkUpdate(mode string) {
-	// Its OK to ignore any errors during getUpdateInfo() here.
-	if older, downloadURL, err := getUpdateInfo(1*time.Second, mode); err == nil {
-		if updateMsg := computeUpdateMessage(downloadURL, older); updateMsg != "" {
+	// Its OK to ignore any errors during doUpdate() here.
+	if updateMsg, _, currentReleaseTime, latestReleaseTime, err := getUpdateInfo(1*time.Second, mode); err == nil {
```
Comment (nitisht) [inline cmd/common-main.go]: 1s timeout seems too little for non-US clients, for example I get a timeout while `getUpdateInfo` checks for update

**Review by nitisht — CHANGES_REQUESTED** [review]

**Review by nitisht — DISMISSED** [review]
LGTM & Tested

**Comment by harshavardhana** [comment]
@nitisht @donatello please review this again.

**Comment by harshavardhana** [comment]
@nitisht @donatello can you review/approve again?

**Review by nitisht — APPROVED** [review]

**Comment by harshavardhana** [comment]
@donatello can you take a look ?

**Review by donatello — APPROVED** [review]

**Comment by deekoder** [comment]
@donatello @nitisht  how about codecoverage and mint for this pr?

**Comment by donatello** [comment]
The code depends on the output of external systems (minio binary download server) - currently due to tight coupling (of external systems) it is not possible to write unit tests to improve the coverage (perhaps a mint test is possible). If we refactor to loosen the coupling I think the code would be more complicated (it was attempted before, and ultimately was not merged).

I think it is ok to merge this ignoring code coverage.

**Comment by deekoder** [comment]
@y4m4 ran mint.so merging.

**Pull request merged**